### ISSUE TITLE
DB Dictionary produces different results than File backed Dictionary for Example.java

### ISSUE THREAD

**Issue by dayre** [extjwnl/extjwnl#15]
When running the file backed dictionary, for both 3.1 and 3.0 the Example.java method "demonstrateSymmetricRelationshipOperation" returns results.

When i import the DB (tried both 3.0 and 3.1) into a MySQL DB, and run the same example using a DB backed Dictionary, the demonstrateSymmetricRelationshipOperation() example returns 0 results and ends up in an IndexOutOfBoundsException exception trying to fetch the first result.  

Any idea why the different behavior from the API using the different Dictionary implementation ?

**Comment by autayeu** [comment]
Thank you for the report. Most likely it's a bug. I have to look, but can't promise to do it quickly. Patches welcome :)

**Comment by dayre** [comment]
I dug a little deeper, it looks to be the order of the synsets for "funny" is different for the file backed vs. the DB dictionary so the example isn't starting with the right synset (1268418 vs. 971199).

I'm not sure this is actually a problem, just that the example makes this assumption.

Filebacked synsets
```
1) [Synset: [Offset: 1268418] [POS: adjective] Words: amusing, comic, comical, funny, laughable, mirthful, risible 
2) [Synset: [Offset: 971199] [POS: adjective] Words: curious, funny, odd, peculiar, queer, rum, rummy, singular
3) [Synset: [Offset: 1923975] [POS: adjective] Words: fishy, funny, shady, suspect, suspicious 
4) [Synset: [Offset: 2555793] [POS: adjective] Words: funny 
```
    
DB backed synsets
```
1) [Synset: [Offset: 971199] [POS: adjective] Words: curious, funny, odd, peculiar, queer, rum, rummy, singular 
2) [Synset: [Offset: 1268418] [POS: adjective] Words: amusing, comic, comical, funny, laughable, mirthful, risible 
3) [Synset: [Offset: 1923975] [POS: adjective] Words: fishy, funny, shady, suspect, suspicious 
4) [Synset: [Offset: 2555793] [POS: adjective] Words: funny 
```

Unless you see the order as a problem, you can likely close this.

**Comment by logological** [comment]
I don't recall the order being specified anywhere, though by the principle of least surprise, I think it would make sense for it to be consistent regardless of whether the backing store is a file or a database.

**Comment by autayeu** [comment]
@dayre Please, can you post a piece of code you've used to produce the listings?

@logological Least surprise (and consistency) is a good principle, but sometimes certain behavior might have to be preserved for backwards compatibility. I'd find the root cause first (time permitting) and then evaluate the consequences of changing the order.

**Comment by dayre** [comment]
I've attached the test i used (rename to Test.java), you need an existing DB wordnet already present then replace the URL params, it runs the file then the db version and produces the following results, note the order of the 971199 synset in each example.

FILE
[Synset: [Offset: 971199] [POS: adjective] ...
[Synset: [Offset: 1268418] [POS: adjective] ...
[Synset: [Offset: 1923975] [POS: adjective] ...
[Synset: [Offset: 2555793] [POS: adjective] ...

DB
[Synset: [Offset: 1268418] [POS: adjective] ...
[Synset: [Offset: 971199] [POS: adjective] ...
[Synset: [Offset: 1923975] [POS: adjective] ...
[Synset: [Offset: 2555793] [POS: adjective] ...

[Test.txt](https://github.com/extjwnl/extjwnl/files/310664/Test.txt)

**Comment by marcosluizfp** [comment]
Hello everybody! 

I evaluated the problem and it seems the cause is a missing information in database: 
- senses of a lemma shall be ranked according to their frequency of occurrence in semantic concordance texts.
In wordnet file, this is done on the lemma line where the senses (synset offsets) are already sorted by rank, but this order was not saved to database (the query returns the senses ordered by the synset offsets int number, not rank).

A proposed solution demands:
a) creation of a column in indexwordsynset to save the lemma rank, for example "rank"
b) change some queries to order the results by "rank"

@dayre I believe you inverted the results on your last post (FILE versus DB).

**Comment by autayeu** [comment]
@marcosluizfp, thank you! That's a great summary! 

The question is "keeping backward compatibility" vs "fixing this". Opinions?

**Comment by marcosluizfp** [comment]
@autayeu, in my opinion, it should be fixed. 

The reason I defend this position is because this is a definition from their authors (Princeton) for the WordNet, so I believe the WordNet community expects this behavior when accessing the dictionary.

**Comment by logological** [comment]
Why not fix it, but add an option for preserving the old, buggy behaviour for those applications which may depend on it?

**Comment by autayeu** [comment]
@logological - reasonable. However, options (and backward compatibility in general) increase the amount of code to support.

extJWNL is a rather stable library, with (I assume) most of its users not having a big problem in recompiling and fixing the code that uses the library. It's also my impression that database-backed dictionary is not particularly popular. But those are assumptions - lack of usage data makes it hard to back them up.

This makes me lean towards fixing it without regard for backward compatibility. Existing users will have to recreate and reimport the database.

By the way, patches are welcome.

**Comment by dayre** [comment]
I've had to revisit this again with another project, so have submitted a pull request containing the proposed changes.     Better late than never !

**Comment by autayeu** [comment]
@dayre Thank you very much! I've merged this and released 1.9.3.

I'm closing the issue, but feel free to reopen should you discover something wrong with this (or release). Thank you.
